snapshots/overlay/overlay.go

```diff
@@ -112,14 +140,26 @@ func NewSnapshotter(root string, opts ...Opt) (snapshots.Snapshotter, error) {
 		logrus.WithError(err).Warnf("cannot detect whether \"userxattr\" option needs to be used, assuming to be %v", userxattr)
 	}
 
-	return &snapshotter{
+	// figure out whether "metacopy" option is recognized by the kernel && needed
+	metacopyState, err := overlayutils.SupportsMetacopy(root)
+	if err != nil {
+		logrus.WithError(err).Warnf("cannot check metacopy support")
+	}
+
+	s := &snapshotter{
 		root:          root,
 		ms:            ms,
 		asyncRemove:   config.asyncRemove,
 		upperdirLabel: config.upperdirLabel,
 		indexOff:      supportsIndex(),
 		userxattr:     userxattr,
-	}, nil
+		useMetacopy:   config.metacopy,
+		metacopyState: metacopyState,
```

Comment:
Yes, we can do like this. In case of CONFIG_OVERLAY_FS_METACOPY=y the metacopy=on option is just ignored and not shown in mount options, the same for CONFIG_OVERLAY_FS_METACOPY=n the metacopy=off. I'll update this place.

Reply:
Oh, sounds great then! :)

Reply:
One note here, I think that we need to keep three states for metacopy field (Enabled, Disabled, Not supported). We need this because of in case of old kernels the mount operation can fail due to metacopy=on/off option is unsupported.

Reply:
First case:
Metacopy supported by kernel and `CONFIG_OVERLAY_FS_METACOPY=Y`. 
In this case overlayfs mounts will be created with metacopy support even if we don't pass `metacopy=on` option to `mount()` system call. 
`config.Metacopy = false`.
As result:
`config.Metacopy = false && metacopySupported = true` == false -> mount with `metacopy=off`-> ok, metacopy disabled for this mount point

Second case:
Metacopy is not supported by kernel. 
`config.Metacopy = false`.
As result:
`config.Metacopy = false && metacopySupported = false` == false -> mount with `metacopy=off` -> mount failed

We need third state to distinguish between these situations.
In your approach if  `CONFIG_OVERLAY_FS_METACOPY=Y` and `config.Metacopy = false` we just don't add `metacopy=on/off` option to mount system call and metacopy will be silently enabled for containers without containerd request.

In case of assumption that nobody set `CONFIG_OVERLAY_FS_METACOPY=Y`, yes, we no need third state.

Reply:
Oh, right, brain fart here (it was implied in the "f metacopy is supported" that I was wrong, as we lost that data combining the bools).

I do think, though, that we don't need the enum state. We need to know if it is supported and if it was enabled, then we can just add the on/off unconditionally. IMHO it seems simpler (less things to know and reason about) and avoids races like the sysfs value changed just between when we read it and when the overlayfs was mounted.

Does it make sense?

Reply:
In fact the first iteration use two booleans, one for on/off and one for supported/unsupported. From my point of view it doesn't matter enum or booleans, for me enum is more clear just maybe because of I am mostly C programmer ). I'll just compare what option looks better and publish better version.
Regarding sysfs, we can just use stat for /sys/modules/overlay/parameters/metacopy to check supported or not and yes, add on/off unconditionally. This should be race-free.